shai/src/main/scala/at/forsyte/apalache/shai/TransExplorerService.scala

```diff
@@ -175,16 +196,16 @@ class TransExplorerService(connections: Ref[Map[UUID, Conn]], parserSemaphore: S
             try { ZIO.succeed(UUID.fromString(conn.id)) }
             catch {
               case _: IllegalArgumentException =>
-                ZIO.fail(
-                    Status.INVALID_ARGUMENT.withDescription("Invalid RPC request: Connection id is not a valid UUID"))
+                val msg = s"Invalid RPC request: Connection id ${conn.id} is not a valid UUID"
+                Log.error(None, msg).andThen(ZIO.fail(Status.INVALID_ARGUMENT.withDescription(msg)))
```

Comment:
This is just sequencing the two effects **in the ZIO concurrency framework**, so it means first log the message "andThen" send the failure. If we bypassed the concurrency framework an just used Scala operations here, we'd lose our guarantees about the order in which the effects are to be executed.

Reply:
Perhaps it helps to note that we could also represent the sequencing of these effects as:

```suggestion
                for {
                  _ <- Log.error(None, msg)
                  _ <- ZIO.fail(Status.INVALID_ARGUMENT.withDescription(msg))
                } yield ()
```

Only this is a bit misleading because we can't ever actually reach the yield. Still, I can change it if you fin this is more helpful.

Reply:
> If we bypassed the concurrency framework an just used Scala operations here, we'd lose our guarantees about the order in which the effects are to be executed by the ZIO scheduler.

I think I share the same basic confusion with Igor, so I'll jump in here.
Aren't we bypassing the ZIO framework anyhow, just one call deeper into lazy logging?

The sequencing here is only necessary because `Log.error` returns a promise, right?

Reply:
>  Aren't we bypassing the ZIO framework anyhow, just one call deeper into LazyLogging?

No. By lifting the effects into ZIO values we are delegating the execution of the effect to the ZIO scheduler. Once the scheduler decides to call function, then the effect is indeed out into the real world, and that's the time for it.

The sequencing is necessary here because we want to ensure that the error message is logged before we send back the response. The fact that the `Log.error` returns a "promise" is what allows us to sequence it within the ZIO schedular.